[shopsys] added new coding standard: forbidden private visibility in namespace

grossmannmartin opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| Enabled and forces not having any private fields and methods in certain namespaces.
|New feature| No <!-- Do not forget to update docs/ -->
|BC breaks| Yes <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| closes #726, closes #664 <!-- Write "closes #123" for the issue to be closed automatically during merge -->
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

According to our coding standards, there should be only public or protected functions, properties, and constants in Entities, Factories, Controllers, Facades, and Repositories in framework package and plugins.

Following namespaces are checked:
- `Shopsys\FrameworkBundle\Model`
- `Shopsys\FrameworkBundle\Component`
- `Shopsys\FrameworkBundle\Controller`
- `Shopsys\FrameworkBundle\DataFixtures\Demo`
- `Shopsys\ProductFeed\GoogleBundle\Model`
- `Shopsys\ProductFeed\HeurekaBundle\Model`
- `Shopsys\ProductFeed\HeurekaDeliveryBundle\Model`
- `Shopsys\ProductFeed\ZboziBundle\Model`

PetrHeinz commented on packages/coding-standards/src/CsFixer/ForbiddenPrivateVisibilityFixer.php:
```diff
@@ -0,0 +1,141 @@
+     */
+    public function configure(array $configuration = null): void
+    {
+        if (array_key_exists(self::OPTION_ANALYZED_NAMESPACE, $configuration) && is_array($configuration[self::OPTION_ANALYZED_NAMESPACE])) {
```
This opens up to misconfiguration, if the array key exists it should always be an array, I would rather throw an exception is such a case.

simara-svatopluk commented:
Good job with the fixer :-)

simara-svatopluk commented:
One last question - who was involved in decision what namespaces have to be `protected` only?

grossmannmartin commented on docs/contributing/coding-standards.md:
```diff
@@ -94,5 +94,10 @@ Besides the rules that are checked by automatic tools, we have few rules for whi
     - Controllers
     - Facades
     - Repositories
+    - whole namespaces 
```
But one of the acceptance criteria is "we have the rule documented" and this seems to me like proper documentation.

simara-svatopluk replied:
You are right.. confusing situation for me. I suggest to contact our product owner because when we have a sniffer/fixer, then we don't document it in "coding standards"

Another issue is that the documentation is located in section "Standards that are not checked automatically" - and that is not true ;-)

grossmannmartin replied:
Yes, I didn't notice it's in that section (Standards that are not checked automatically).
It doesn't make sense this way. For now, I removed the note and I will discuss it with product owner.

grossmannmartin commented:
> One last question - who was involved in decision what namespaces have to be `protected` only?

I started from task definition in user story (= `Shopsys\FrameworkBundle\Model` and `Shopsys\FrameworkBundle\Component`), I believe it was consulted internally. 

Then I added DataFixtures (because without them, we have to have exception for ie. AbstractReferenceFixture in Components)

And I added Model namespace from product-feed-* packages thanks to #664 and because it makes sense to me like this (and this rule is already applied there)

grossmannmartin commented on packages/coding-standards/src/CsFixer/ForbiddenPrivateVisibilityFixer.php:
```diff
@@ -0,0 +1,141 @@
+     */
+    public function isCandidate(Tokens $tokens): bool
+    {
+        return $tokens->isAllTokenKindsFound([T_CLASS, T_PRIVATE]) && $this->checkNamespace($tokens);
```
No, we don't need to 👍

grossmannmartin commented on packages/coding-standards/easy-coding-standard.yml:
```diff
@@ -18,6 +18,12 @@ services:
     Shopsys\CodingStandards\CsFixer\ForbiddenDumpFixer: ~
+    Shopsys\CodingStandards\CsFixer\ForbiddenPrivateVisibilityFixer:
+        analyzed_namespaces:
+            - Shopsys\FrameworkBundle\Model
```
You're right. I moved it, but in monorepo have to be still a whole list of namespaces. But there I believe it's ok.

PetrHeinz commented on docs/contributing/coding-standards.md:
```diff
@@ -94,5 +94,10 @@ Besides the rules that are checked by automatic tools, we have few rules for whi
+        - `Shopsys\FrameworkBundle\Model`
+        - `Shopsys\FrameworkBundle\Component`
+        - `Shopsys\FrameworkBundle\Controller`
+        - `Shopsys\FrameworkBundle\DataFixtures\Demo`
```
This whole section is under the heading *Standards that are not checked automatically*. As it is checked automatically, I would rather remove this bullet altogether (starting by `- Visibility of all properties, methods and constants of`).

grossmannmartin replied:
Right, I didn't notice that. As I said in https://github.com/shopsys/shopsys/pull/788#discussion_r254157257, I deleted it and I will discuss this topic with product owner.

simara-svatopluk commented on packages/coding-standards/easy-coding-standard.yml:
```diff
@@ -18,6 +18,7 @@ services:
 
     # Shopsys Checkers
     Shopsys\CodingStandards\CsFixer\ForbiddenDumpFixer: ~
+    Shopsys\CodingStandards\CsFixer\ForbiddenPrivateVisibilityFixer: ~
```
Yep, the last nitpick note - if I understand correctly, the Fixer does nothing because it is not configured, right? If so, I would prefer to remove this line

grossmannmartin replied:
You're right. Removed.

simara-svatopluk commented:
@grossmannmartin Please review commit messages - 2. and 3. commits have wrong formatting in sub bullets.
Also please add note about removed documentation to 2. commit

I'll keep testing :-)

simara-svatopluk commented on packages/product-feed-google/easy-coding-standard.yml:
```diff
@@ -1,6 +1,15 @@
     - { resource: '%vendor_dir%/shopsys/coding-standards/easy-coding-standard.yml', ignore_errors: true  }
 
+services:
+    forbidden_private_visibility_fixer.google:
```
I would like to have here a full package name to set some kind of standard (also in other product-feed-* packages)

grossmannmartin replied:
seems reasonable 👍

simara-svatopluk commented:
Reviewed again, tested. Ok!

simara-svatopluk commented:
Reviewed again, tested, works!

simara-svatopluk commented:
Reviewed again, tested, works!

I have only one comment left - naming private services.
